src/lib/createAndHandleSegments.ts

```diff
@@ -0,0 +1,180 @@
+
+  const rect = timelineContainer && timelineContainer?.getBoundingClientRect();
+  if (rect) {
+    const percent =
+      Math.min(Math.max(0, e?.x - rect?.x), rect?.width) / rect?.width;
+    return percent;
+  }
+}
+function changeCurrentTimeOnClick(e: any, player: any) {
+  const timelineContainer = document.querySelector('#timeline-container');
+
+  const rect = timelineContainer && timelineContainer?.getBoundingClientRect();
+  let percent;
+  if (rect) {
+    percent = Math.min(Math.max(0, e?.x - rect?.x), rect?.width) / rect?.width;
+    player?.currentTime(percent * player?.duration());
+  }
+}
+
+function setCurrentLabel(
+  e: any,
+  preview_position: any,
+  segments: Segment[],
+  player: any,
+) {
+  let currentLabel = segments[0].title;
+
+  const currentPos = preview_position * player?.duration();
+  if (preview_position !== 0) {
+    segments?.forEach((each: Segment) => {
+      if (each?.start < currentPos && each?.end > currentPos) {
+        currentLabel = each?.title;
+      }
+    });
+  }
+
+  const timelineLabel = document.querySelector(
+    '#timeline-label',
+  ) as HTMLDivElement;
+  const timelineContainer = document.querySelector(
+    '#timeline-container',
+  ) as HTMLDivElement;
+
+  const rect = timelineContainer && timelineContainer?.getBoundingClientRect();
+
+  let timelineLabelWidth = timelineLabel?.offsetWidth / 2;
+
+  if (timelineLabel && rect && timelineContainer) {
+    timelineLabel.innerHTML = `<div>${formatTime(preview_position.toFixed(2) * player?.duration())} - ${currentLabel}</div>`;
+
+    if (Math.max(0, e?.x - rect?.x) < timelineLabel?.offsetWidth / 2) {
+      timelineLabelWidth = Math.max(0, e?.x - rect?.x);
+    } else if (
+      Math.max(0, e?.x - rect?.x) >
+      timelineContainer.offsetWidth - timelineLabel?.offsetWidth / 2
+    ) {
+      timelineLabelWidth =
+        timelineLabel?.offsetWidth -
+        (timelineContainer?.offsetWidth - (e?.x - rect?.x));
+    }
+
+    timelineLabel.style.left = `calc(${preview_position * 100}% - ${timelineLabelWidth}px)`;
+  }
+}
+
+let isMouseOver = false;
+let isMouseDown = false;
+
+export function updateTimeline(e: any, player: any, segments: Segment[]) {
+  let preview_position = 0;
+
+  if (e.type === 'timeupdate') {
+    if (!isMouseDown) {
+      progessTimeline(segments, player);
+    }
+
+    if (!isMouseOver) {
+      preview_position = previewViaBuffer(player);
+    }
+  }
+  if (e.type === 'click') {
+    changeCurrentTimeOnClick(e, player);
+    isMouseDown = false;
+  }
+
+  if (e.type === 'mouseover') {
+    isMouseOver = true;
+
+    preview_position = previewViaMouseOverOrMove(e) as number;
+  }
+  if (e.type === 'mousemove') {
+    isMouseOver = true;
+
+    preview_position = previewViaMouseOverOrMove(e) as number;
+
+    setCurrentLabel(e, preview_position, segments, player);
+
+    if (isMouseDown) {
+      changeCurrentTimeOnClick(e, player);
+      progessTimeline(segments, player);
+    }
```

Comment:
this code handles preview via buffer and currentTime of player based on mouse events on timline-container to rovide youtube like experience